File: src/openjarvis/engine/ollama.py
"""Ollama inference engine backend."""

from __future__ import annotations

import json
import logging
import os
from collections.abc import AsyncIterator, Sequence
from typing import Any, Dict, List

import httpx

from openjarvis.core.registry import EngineRegistry
from openjarvis.core.types import Message
from openjarvis.engine._base import (
    EngineConnectionError,
    InferenceEngine,
    messages_to_dicts,
)

logger = logging.getLogger(__name__)


@EngineRegistry.register("ollama")
class OllamaEngine(InferenceEngine):
    """Ollama backend via its native HTTP API."""

    engine_id = "ollama"

    _DEFAULT_HOST = "http://localhost:11434"

    def __init__(
        self,
        host: str | None = None,
        *,
        timeout: float = 1800.0,
    ) -> None:
        # Priority: explicit host (from config.toml) > OLLAMA_HOST env var > default
        if host is None:
            env_host = os.environ.get("OLLAMA_HOST")
            host = env_host or self._DEFAULT_HOST
        self._host = host.rstrip("/")
        self._client = httpx.Client(base_url=self._host, timeout=timeout)

    def generate(
        self,
        messages: Sequence[Message],
        *,
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> Dict[str, Any]:
        msg_dicts = messages_to_dicts(messages)
        # Ollama expects tool_call arguments as dicts, not JSON strings
        for md in msg_dicts:
            for tc in md.get("tool_calls", []):
                fn = tc.get("function", {})
                args = fn.get("arguments")
                if isinstance(args, str):
                    try:
                        fn["arguments"] = json.loads(args)
                    except (json.JSONDecodeError, TypeError):
                        pass
        payload: Dict[str, Any] = {
            "model": model,
            "messages": msg_dicts,
            "stream": False,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
                "num_ctx": kwargs.get("num_ctx", 8192),
            },
        }
        # Pass tools if provided
        tools = kwargs.get("tools")
        if tools:
            payload["tools"] = tools

        # Apply structured output / JSON mode
        response_format = kwargs.get("response_format")
        if response_format is not None:
            from openjarvis.engine._stubs import ResponseFormat

            if isinstance(response_format, ResponseFormat):
                payload["format"] = "json"
            elif isinstance(response_format, dict):
                payload["format"] = "json"
        try:
            resp = self._client.post("/api/chat", json=payload)
            if resp.status_code == 400 and tools:
                # Model may not support function calling -- retry without tools
                payload.pop("tools", None)
                resp = self._client.post("/api/chat", json=payload)
            resp.raise_for_status()
        except (httpx.ConnectError, httpx.TimeoutException) as exc:
            raise EngineConnectionError(
                f"Ollama not reachable at {self._host}"
            ) from exc
        except httpx.HTTPStatusError as exc:
            body = exc.response.text[:500] if exc.response else ""
            raise RuntimeError(
                f"Ollama returned {exc.response.status_code}: {body}"
            ) from exc
        data = resp.json()
        prompt_tokens = data.get("prompt_eval_count", 0)
        completion_tokens = data.get("eval_count", 0)
        content = data.get("message", {}).get("content", "")
        result: Dict[str, Any] = {
            "content": content,
            "usage": {
                "prompt_tokens": prompt_tokens,
                "completion_tokens": completion_tokens,
                "total_tokens": prompt_tokens + completion_tokens,
            },
            "model": data.get("model", model),
            "finish_reason": "stop",
        }
        # Extract timing from Ollama response (nanoseconds → seconds)
        result["ttft"] = data.get("prompt_eval_duration", 0) / 1e9
        result["engine_timing"] = {k: data[k] for k in
            ("total_duration", "load_duration", "prompt_eval_duration", "eval_duration")
            if k in data}
        # Extract tool calls if present
        raw_tool_calls = data.get("message", {}).get("tool_calls", [])
        if raw_tool_calls:
            tool_calls = []
            for i, tc in enumerate(raw_tool_calls):
                raw_args = tc.get("function", {}).get(
                    "arguments", "{}",
                )
                tool_calls.append({
                    "id": tc.get("id", f"call_{i}"),
                    "name": tc.get("function", {}).get("name", ""),
                    "arguments": (
                        json.dumps(raw_args)
                        if isinstance(raw_args, dict)
                        else raw_args
                    ),
                })
            result["tool_calls"] = tool_calls
        return result

    async def stream(
        self,
        messages: Sequence[Message],
        *,
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> AsyncIterator[str]:
        payload: Dict[str, Any] = {
            "model": model,
            "messages": messages_to_dicts(messages),
            "stream": True,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
                "num_ctx": kwargs.get("num_ctx", 8192),
            },
        }
        try:
            with self._client.stream("POST", "/api/chat", json=payload) as resp:
                resp.raise_for_status()
                for line in resp.iter_lines():
                    if not line.strip():
                        continue
                    try:
                        chunk = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    content = chunk.get("message", {}).get("content", "")
                    if content:
                        yield content
                    if chunk.get("done", False):
                        break
        except (httpx.ConnectError, httpx.TimeoutException) as exc:
            raise EngineConnectionError(
                f"Ollama not reachable at {self._host}"
            ) from exc

    def list_models(self) -> List[str]:
        try:
            resp = self._client.get("/api/tags")
            resp.raise_for_status()
        except (
            httpx.ConnectError, httpx.TimeoutException, httpx.HTTPStatusError,
        ) as exc:
            logger.warning(
                "Failed to list models from Ollama at %s: %s",
                self._host, exc,
            )
            return []
        data = resp.json()
        return [m["name"] for m in data.get("models", [])]

    def health(self) -> bool:
        try:
            resp = self._client.get("/api/tags", timeout=2.0)
            return resp.status_code == 200
        except Exception as exc:
            logger.debug("Ollama health check failed at %s: %s", self._host, exc)
            return False

    def close(self) -> None:
        self._client.close()


__all__ = ["OllamaEngine"]

File: repos/openjarvis/rust/crates/openjarvis-engine/src/ollama.rs
//! Ollama inference engine backend.

use crate::traits::{InferenceEngine, TokenStream};
use openjarvis_core::error::{EngineError, OpenJarvisError};
use openjarvis_core::{GenerateResult, Message, ToolCall, Usage};
use serde_json::Value;

/// Ollama backend via its native HTTP API.
pub struct OllamaEngine {
    host: String,
    client: reqwest::blocking::Client,
    timeout: std::time::Duration,
}

impl OllamaEngine {
    pub fn new(host: &str, timeout_secs: f64) -> Self {
        let host = host.trim_end_matches('/').to_string();
        let timeout = std::time::Duration::from_secs_f64(timeout_secs);
        let client = reqwest::blocking::Client::builder()
            .timeout(timeout)
            .build()
            .unwrap_or_default();
        Self {
            host,
            client,
            timeout,
        }
    }

    pub fn with_defaults() -> Self {
        Self::new("http://localhost:11434", 120.0)
    }
}

impl Default for OllamaEngine {
    fn default() -> Self {
        Self::with_defaults()
    }
}

#[async_trait::async_trait]
impl InferenceEngine for OllamaEngine {
    fn engine_id(&self) -> &str {
        "ollama"
    }

    fn generate(
        &self,
        messages: &[Message],
        model: &str,
        temperature: f64,
        max_tokens: i64,
        extra: Option<&Value>,
    ) -> Result<GenerateResult, OpenJarvisError> {
        let msg_dicts = crate::traits::messages_to_dicts(messages);
        let mut payload = serde_json::json!({
            "model": model,
            "messages": msg_dicts,
            "stream": false,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
            }
        });

        if let Some(extra_val) = extra {
            if let Some(tools) = extra_val.get("tools") {
                payload["tools"] = tools.clone();
            }
        }

        let resp = self
            .client
            .post(format!("{}/api/chat", self.host))
            .json(&payload)
            .send()
            .map_err(|e| {
                OpenJarvisError::Engine(EngineError::Connection(format!(
                    "Ollama not reachable at {}: {}",
                    self.host, e
                )))
            })?;

        if !resp.status().is_success() {
            return Err(OpenJarvisError::Engine(EngineError::Http(format!(
                "Ollama returned status {}",
                resp.status()
            ))));
        }

        let data: Value = resp.json().map_err(|e| {
            OpenJarvisError::Engine(EngineError::Deserialization(e.to_string()))
        })?;

        let prompt_tokens = data["prompt_eval_count"].as_i64().unwrap_or(0);
        let completion_tokens = data["eval_count"].as_i64().unwrap_or(0);
        let content = data["message"]["content"]
            .as_str()
            .unwrap_or("")
            .to_string();
        let model_name = data["model"].as_str().unwrap_or(model).to_string();
        let ttft = data["prompt_eval_duration"].as_f64().unwrap_or(0.0) / 1e9;

        let tool_calls = data["message"]["tool_calls"]
            .as_array()
            .map(|arr| {
                arr.iter()
                    .enumerate()
                    .map(|(i, tc)| {
                        let func = &tc["function"];
                        let args = if func["arguments"].is_object() {
                            serde_json::to_string(&func["arguments"]).unwrap_or_default()
                        } else {
                            func["arguments"]
                                .as_str()
                                .unwrap_or("{}")
                                .to_string()
                        };
                        ToolCall {
                            id: tc["id"]
                                .as_str()
                                .unwrap_or(&format!("call_{}", i))
                                .to_string(),
                            name: func["name"].as_str().unwrap_or("").to_string(),
                            arguments: args,
                        }
                    })
                    .collect()
            });

        Ok(GenerateResult {
            content,
            usage: Usage {
                prompt_tokens,
                completion_tokens,
                total_tokens: prompt_tokens + completion_tokens,
            },
            model: model_name,
            finish_reason: "stop".into(),
            tool_calls,
            ttft,
            cost_usd: 0.0,
            metadata: std::collections::HashMap::new(),
        })
    }

    async fn stream(
        &self,
        messages: &[Message],
        model: &str,
        temperature: f64,
        max_tokens: i64,
        _extra: Option<&Value>,
    ) -> Result<TokenStream, OpenJarvisError> {
        let msg_dicts = crate::traits::messages_to_dicts(messages);
        let payload = serde_json::json!({
            "model": model,
            "messages": msg_dicts,
            "stream": true,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
            }
        });

        let async_client = reqwest::Client::builder()
            .timeout(self.timeout)
            .build()
            .map_err(|e| {
                OpenJarvisError::Engine(EngineError::Connection(e.to_string()))
            })?;

        let resp = async_client
            .post(format!("{}/api/chat", self.host))
            .json(&payload)
            .send()
            .await
            .map_err(|e| {
                OpenJarvisError::Engine(EngineError::Connection(format!(
                    "Ollama not reachable at {}: {}",
                    self.host, e
                )))
            })?;

        if !resp.status().is_success() {
            return Err(OpenJarvisError::Engine(EngineError::Http(format!(
                "Ollama returned status {}",
                resp.status()
            ))));
        }

        let byte_stream = resp.bytes_stream();
        use futures::StreamExt;

        let token_stream = byte_stream.filter_map(|chunk_result| async {
            match chunk_result {
                Ok(bytes) => {
                    let text = String::from_utf8_lossy(&bytes);
                    for line in text.lines() {
                        if line.trim().is_empty() {
                            continue;
                        }
                        if let Ok(chunk) = serde_json::from_str::<Value>(line) {
                            let content = chunk["message"]["content"]
                                .as_str()
                                .unwrap_or("")
                                .to_string();
                            if !content.is_empty() {
                                return Some(Ok(content));
                            }
                            if chunk["done"].as_bool().unwrap_or(false) {
                                return None;
                            }
                        }
                    }
                    None
                }
                Err(e) => Some(Err(OpenJarvisError::Engine(EngineError::Streaming(
                    e.to_string(),
                )))),
            }
        });

        Ok(Box::pin(token_stream))
    }

    fn list_models(&self) -> Result<Vec<String>, OpenJarvisError> {
        let resp = self
            .client
            .get(format!("{}/api/tags", self.host))
            .send()
            .map_err(|_| {
                OpenJarvisError::Engine(EngineError::Connection(
                    "Ollama not reachable".into(),
                ))
            })?;

        if !resp.status().is_success() {
            return Ok(vec![]);
        }

        let data: Value = resp.json().unwrap_or(Value::Null);
        let models = data["models"]
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|m| m["name"].as_str().map(String::from))
                    .collect()
            })
            .unwrap_or_default();
        Ok(models)
    }

    fn health(&self) -> bool {
        self.client
            .get(format!("{}/api/tags", self.host))
            .timeout(std::time::Duration::from_secs(2))
            .send()
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ollama_default_host() {
        let engine = OllamaEngine::with_defaults();
        assert_eq!(engine.engine_id(), "ollama");
        assert_eq!(engine.host, "http://localhost:11434");
    }
}
